src/olympia/api/tests/test_authentication.py

```diff
@@ -0,0 +1,231 @@
+    def test_decode_decoding_error(self, jwt_decode_handler):
+        jwt_decode_handler.side_effect = jwt.DecodeError
+        with self.assertRaises(AuthenticationFailed) as ctx:
+            self.auth.authenticate(self.request('whatever'))
+
+        assert ctx.exception.detail == 'Error decoding signature.'
+
+    @mock.patch('olympia.api.jwt_auth.jwt_decode_handler')
+    def test_decode_invalid_token(self, jwt_decode_handler):
+        jwt_decode_handler.side_effect = jwt.InvalidTokenError
+        with self.assertRaises(AuthenticationFailed) as ctx:
+            self.auth.authenticate(self.request('whatever'))
+
+        assert ctx.exception.detail == 'Invalid JWT Token.'
+
+    def test_refuse_refreshable_tokens(self):
+        # We should not accept tokens that have an orig_iat field set.
+        api_key = self.create_api_key(self.user)
+        payload = self.auth_token_payload(self.user, api_key.key)
+        payload['orig_iat'] = timegm(payload['iat'].utctimetuple())
+        token = self.encode_token_payload(payload, api_key.secret)
+
+        with self.assertRaises(AuthenticationFailed) as ctx:
+            self.auth.authenticate(self.request(token))
+
+        assert ctx.exception.detail == (
+            "API key based tokens are not refreshable, don't include "
+            "`orig_iat` in their payload.")
+
+    def test_cant_refresh_token(self):
+        # Developers generate tokens, not us, they should not be refreshable,
+        # the refresh implementation does not even know how to decode them.
+        api_key = self.create_api_key(self.user)
+        payload = self.auth_token_payload(self.user, api_key.key)
+        payload['orig_iat'] = timegm(payload['iat'].utctimetuple())
+        token = self.encode_token_payload(payload, api_key.secret)
+
+        refresh_token_url = reverse('frontend-token-refresh')
+        response = self.client.post(refresh_token_url, data={'token': token})
+        assert response.status_code == 400
+        data = json.loads(response.content)
+        assert data == {'non_field_errors': ['Error decoding signature.']}
+
+
+class TestJWTKeyAuthProtectedView(WithDynamicEndpoints, JWTAuthKeyTester):
+    fixtures = ['base/addon_3615']
+
+    def setUp(self):
+        super(TestJWTKeyAuthProtectedView, self).setUp()
+        self.endpoint(JWTKeyAuthTestView)
+        self.client.logout_api()  # just to be sure!
+        self.user = UserProfile.objects.get(email='del@icio.us')
+
+    def request(self, method, *args, **kw):
+        handler = getattr(self.client, method)
+        return handler('/en-US/firefox/dynamic-endpoint', *args, **kw)
+
+    def jwt_request(self, token, method, *args, **kw):
+        return self.request(method,
+                            HTTP_AUTHORIZATION='JWT {}'.format(token),
+                            *args, **kw)
+
+    def test_get_requires_auth(self):
+        res = self.request('get')
+        assert res.status_code == 401, res.content
```

Comment:
Here I only care whether the auth class is applied correctly or not, I don't care about the verb since it's a test view anyway.